### PR TITLE
[IMP] website_forum: improve filters in forum

### PR DESCRIPTION
**Pull request by angv-odoo**
This commit improves filters in forum so that moderator can see all non-deleted posts on main page, use filters to move to flagged, closed or pending post lists, as well as see deleted posts and reopen them through filters if need arises.

task-3349373

### CODE REVIEW

File: addons/website_forum/views/forum_forum_templates_moderation.xml [hunk 1]
```diff
@@ -151,9 +151,9 @@
                             t-out="question.name"/>
                     </div>
                     <div class="col-lg-3 d-lg-flex">
-                        <span t-if="not question.active and question.state=='offensive'" class="badge bg-warning text-wrap">Offensive</span>
-                        <span t-if="not question.active and question.state=='offensive' and question.closed_reason_id" class="badge bg-warning text-wrap" t-out="question.closed_reason_id.name.capitalize()"/>
-                        <span t-if="question.closed_reason_id" class="badge bg-danger text-wrap" t-out="question.closed_reason_id.name.capitalize()"/>
+                        <span t-if="question.state=='offensive'" class="badge bg-warning text-wrap">Offensive</span>
+                        <span t-if="question.state=='offensive' and question.closed_reason_id" class="badge bg-warning text-wrap" t-out="question.closed_reason_id.name[0].upper() + question.closed_reason_id.name[1:]"/>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_moderation.xml]: why this change ?

File: addons/website_forum/views/forum_forum_templates_moderation.xml [hunk 2]
```diff
@@ -151,9 +151,9 @@
...
+                        <span t-if="question.closed_reason_id" class="badge bg-danger text-wrap" t-out="question.closed_reason_id.name[0].upper() + question.closed_reason_id.name[1:]"/>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_moderation.xml]: same here

File: addons/website_forum/models/forum_post.py [hunk 3]
```diff
@@ -808,10 +806,12 @@ def _search_get_detail(self, website, order, options):
             'name': {'name': 'name', 'type': 'text', 'match': True},
             'website_url': {'name': 'website_url', 'type': 'text', 'truncate': False},
         }
-
+        values = options.get('values') or False
         domain = website.website_domain()
-        domain += [('parent_id', '=', False), ('state', '=', 'active'), ('can_view', '=', True)]
+        domain += [('parent_id', '=', False), ('can_view', '=', True)]
         forum = options.get('forum')
+        if values == False or values.get('user').karma < values.get('forum').karma_moderate:
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py]:
```
>>> 0 == False
True
```

you shouls do `values is False`

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 4]
```diff
@@ -47,11 +47,11 @@
...
-                        <span t-elif="not question.active" class="badge bg-danger">
-                            <t t-if="question.state!='offensive'">Deleted</t>
-                            <t t-if="question.state=='offensive'">Offensive</t>
-                            <t t-if="question.state=='offensive' and question.closed_reason_id">
-                                <t t-out="question.closed_reason_id.name.capitalize()"/>
+                        <span t-elif="not question.active" class="badge bg-danger">Deleted
+                        </span>
+                        <span t-elif="question.state == 'offensive'" class="badge bg-danger">Offensive
+                            <t t-if="question.closed_reason_id">
+                                <t t-out="question.closed_reason_id.name[0].upper() + question.closed_reason_id.name[1:]"/>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml]: why this change ? :/

**Review by std-odoo — COMMENTED** [review]
@angv-odoo Bloup :)

File: addons/website_forum/views/forum_forum_templates_moderation.xml (hunk 1 above)
Comment (angv-odoo) [reply addons/website_forum/views/forum_forum_templates_moderation.xml]: if question state is offensive it is already not active, so there is no need for the check. as for capitalize(), it was probably changed in previous commits I missed it during rebase, corrected every occurrence now :)

Comment (std-odoo) [reply addons/website_forum/views/forum_forum_templates_moderation.xml]: yes it was about `capitalize` :)

Comment (angv-odoo) [reply addons/website_forum/views/forum_forum_templates_moderation.xml]: I corrected that part :D Could you look at it again and tell me if there is anything else I can try to improve?

Comment (std-odoo) [reply addons/website_forum/views/forum_forum_templates_moderation.xml]: no, that's perfect, thanks :)

File: addons/website_forum/models/forum_post.py [hunk 5]
```diff
@@ -554,8 +554,6 @@ def flag(self):
         for post in self:
             if not post.can_flag:
                 raise AccessError(_('%d karma required to flag a post.', post.forum_id.karma_flag))
-            if post.state == 'flagged':
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:557]: shouldn't you remove the error from the JS ?

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 6]
```diff
@@ -39,10 +39,10 @@
                     <h3 t-attf-class="col-lg-10 mb-0" t-out="question.name"/>
                     <div class="col d-flex justify-content-end align-items-center">
                         <i t-if="question.state == 'close'" class="fa fa-lock ms-2 fs-4" title="Closed" data-bs-toggle="tooltip" data-bs-placement="top"/>
-                        <span t-elif="not question.active" class="badge bg-danger">
-                            <t t-if="question.state!='offensive'">Deleted</t>
-                            <t t-if="question.state=='offensive'">Offensive</t>
-                            <t t-if="question.state=='offensive' and question.closed_reason_id">
+                        <span t-elif="not question.active" class="badge bg-danger">Deleted
+                        </span>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml]: can be inlined

**Review by std-odoo — COMMENTED** [review]
@angv-odoo Hi :) 

2 more comments, otherwise LGTM :)

File: addons/website_forum/models/forum_post.py (hunk 5 above)
Comment (angv-odoo) [reply addons/website_forum/models/forum_post.py:557]: yes it should, corrected it now :)

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 7]
```diff
@@ -140,16 +139,28 @@
                     </t>
                     <t t-else="">
                         <t t-if="forum.mode == 'questions'">
-                            <a t-attf-href="?#{ keep_query('search', 'sorting', 'my', filters='solved') }"
-                               class="dropdown-item">Solved
+                            <a t-attf-href="/forum/#{slug(forum)}?filters=solved"
+                                class="dropdown-item">Solved
```
Comment (flch-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml]: Why ? (sorry if it's on the other PR)

File: addons/website_forum/views/forum_forum_templates_post.xml [hunk 8]
```diff
@@ -37,6 +37,11 @@
         <!-- [TODO] Note: after closing and reopening a post, the closed_reason_id remains on post  -->
         <!-- <t t-out="question.closed_reason_id.name"></t> -->
         <!-- [TODO] non active posts aren't listed -->
```
Comment (flch-odoo) [inline addons/website_forum/views/forum_forum_templates_post.xml]: Is there something you can do about these? Either solve or remove if outdated?

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 9]
```diff
@@ -39,10 +39,9 @@
                     <h3 t-attf-class="col-lg-10 mb-0" t-out="question.name"/>
                     <div class="col d-flex justify-content-end align-items-center">
                         <i t-if="question.state == 'close'" class="fa fa-lock ms-2 fs-4" title="Closed" data-bs-toggle="tooltip" data-bs-placement="top"/>
-                        <span t-elif="not question.active" class="badge bg-danger">
-                            <t t-if="question.state!='offensive'">Deleted</t>
-                            <t t-if="question.state=='offensive'">Offensive</t>
-                            <t t-if="question.state=='offensive' and question.closed_reason_id">
+                        <span t-elif="not question.active" class="badge bg-danger">Deleted</span>
+                        <span t-elif="question.state == 'offensive'" class="badge bg-danger">Offensive
```
Comment (flch-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml:43]: don't we need a margin between the two here? Just guessing at this point

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 10]
```diff
@@ -118,7 +117,7 @@
                     </t>
                 </a>
                 <div class="dropdown-menu" role="menu">
-                    <a t-attf-href="?#{ keep_query('search', 'sorting', filters='all') }"
+                    <a t-attf-href="/forum/#{slug(forum)}"
```
Comment (flch-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml]: Also here

File: addons/website_forum/views/forum_forum_templates_post.xml [hunk 11]
```diff
@@ -34,9 +34,11 @@
...
-        <!-- <t t-out="question.closed_reason_id.name"></t> -->
-        <!-- [TODO] non active posts aren't listed -->
+        <span t-if="question.active" class="text-muted">
+                <t t-if="question.state=='pending'"> [Waiting Validation]</t>
+                <t t-if="question.state=='flagged' and user.karma > forum.karma_flag"> [Flagged]</t>
+                <t t-if="question.state=='close'"> [Closed]</t>
+        </span>
         <span t-if="not question.active" class="text-muted">
             <t t-if="question.state!='offensive'"> [Deleted]</t>
             <t t-if="question.state=='offensive'"> [Offensive]</t>
```
Comment (flch-odoo) [inline addons/website_forum/views/forum_forum_templates_post.xml]:
```suggestion
        <span class="text-muted">
                <t t-if="question.state=='pending'"> [Waiting Validation]</t>
                <t t-if="question.state=='flagged' and user.karma > forum.karma_flag"> [Flagged]</t>
                <t t-if="question.state=='close'"> [Closed]</t>
                <t t-if="not question.active"> [Deleted]</t>
                <t t-if="question.state=='offensive'"> [Offensive]</t>
```

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 12]
```diff
@@ -141,15 +140,27 @@
...
+                                class="dropdown-item">Unsolved
                             </a>
                         </t>
                         <a t-attf-href="?#{ keep_query('search', 'sorting', 'my', filters='unanswered') }"
-                           class="dropdown-item">Unanswered
+                            class="dropdown-item">Unanswered
                         </a>
+                        <t t-if="user.karma>=forum.karma_moderate">
+                            <div class="dropdown-divider"/>
+                            <a t-attf-href="/forum/#{slug(forum)}/validation_queue"
```
Comment (flch-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml]: It's a detail, but could you keep please the spaces in `#{ something }` as we did in this file?

**Review by flch-odoo — COMMENTED** [review]
A few more thoughts :)

File: addons/website_forum/views/forum_forum_templates_post.xml (hunk 11 above)
Comment (angv-odoo) [reply addons/website_forum/views/forum_forum_templates_post.xml]: In case I change it the way you proposed, we will get multiple "tags" for each post. Like for example if the post is closed and then someone decides to delete the post, the deleted post will have both [Closed] and [Deleted] next to its title. Do you think it's worth showing the state of the post when the post is already deleted?

File: addons/website_forum/views/forum_forum_templates_layout.xml (hunk 9 above)
Comment (angv-odoo) [reply addons/website_forum/views/forum_forum_templates_layout.xml:43]: they will not show together ever because of t-elif, but if we decide that we want to show both badges at the same time in case of deleted offensive posts(the reason why I don't think it's necessary is on comment below), then yes, we will need a few pixel margin on left :D

File: addons/website_forum/views/forum_forum_templates_post.xml (hunk 11 above)
Comment (flch-odoo) [reply addons/website_forum/views/forum_forum_templates_post.xml]: Sorry, I mean't `t-elif`, and to be optimized. If we don't want multiple tags, start with the active test. Let's not duplicate for now if it isn't already.

File: addons/website_forum/static/src/js/website_forum.js [hunk 13]
```diff
@@ -342,11 +340,12 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
                     $(elem).nextAll('.flag_validator').removeClass('d-none');
                 } else if (data.success === 'post_flagged_non_moderator') {
-                    const forumAnswer = elem.closest('.forum_answer');
-                    elem.innerText = _t(' Flagged');
-                    elem.prepend(child);
-                    forumAnswer.fadeIn(1000);
-                    forumAnswer.slideUp(1000);
+                    this.displayNotification({
+                        message: _t("Moderator has been notified about this post."),
+                        title: _t("Post has been flagged"),
```
Comment (flch-odoo) [inline addons/website_forum/static/src/js/website_forum.js]: "Moderators will be notified about this post." ?

"The post has been successfully flagged" `or` "Post successfully flagged" `or` "The post was just flagged" `or` "Post flagged" look better than the current proposition, if I may raise this point.

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 14]
```diff
@@ -141,15 +140,27 @@
                     <t t-else="">
                         <t t-if="forum.mode == 'questions'">
                             <a t-attf-href="?#{ keep_query('search', 'sorting', 'my', filters='solved') }"
-                               class="dropdown-item">Solved
+                                class="dropdown-item">Solved
```
Comment (flch-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml]: The comment about this seems to have been lost. I intently aligned on the t of t-attf-href ;) .

**Review by flch-odoo — COMMENTED** [review]
And the final 2 ;-)
